src/components/TableDialog.vue

```diff
@@ -157,6 +187,88 @@ export default Vue.extend({
         this.fileName = this.fileName || getFileName(file);
         this.fileUploadError = null;
       }
+
+      interface TypeScore {
+        strings: Set<string>;
+        number: number;
+        date: number;
+        total: number;
+      }
+      const columnTypes = new Map<string, TypeScore>();
+
+      Papa.parse(file, {
+        header: true,
+        skipEmptyLines: true,
+        step(row: ParseResult<{}>) {
+          // eslint-disable-next-line @typescript-eslint/no-explicit-any
+          const data = row.data as { [key: string]: any };
+
+          Object.keys(data).forEach((key: string) => {
+            if (!columnTypes.has(key)) {
+              columnTypes.set(key, {
+                strings: new Set<string>(),
+                number: 0,
+                date: 0,
+                total: 0,
+              });
+            }
+
+            const entry = columnTypes.get(key);
+            if (entry === undefined) {
+              throw new Error('impossible');
+            }
+
+            // Pass the value through a gauntlet of heuristics to see what types
+            // it can convert to.
+            const value = data[key];
+            entry.total += 1;
+
+            // Test for categoriness by counting up all the unique strings in
+            // the column.
+            entry.strings.add(value);
+
+            // See if the value can be converted to a number. Empty strings
+            // convert to 0, so those are excluded specifically.
+            if (value !== '' && !Number.isNaN(Number(value))) {
+              entry.number += 1;
+            }
+
+            // See if the value looks like a date.
+            if (dayjs(value).isValid()) {
+              entry.date += 1;
+            }
+          });
+        },
+
+        complete: () => {
+          const typeRecs = new Map<string, MultinetType>();
+
+          columnTypes.forEach((entry, field) => {
+            const isKey = field === '_key' || field === '_from' || field === '_to';
+            const category = entry.strings.size <= 10;
+            const number = entry.number === entry.total;
+            const date = entry.date === entry.total;
+
+            let rec: MultinetType = 'label';
+            if (isKey) {
+              rec = 'label';
+            } else if (category && !number && !date) {
+              rec = 'category';
+            } else if (number) {
+              rec = 'number';
+            } else if (date) {
+              rec = 'date';
+            }
+
+            typeRecs.set(field, rec);
+          });
+
+          this.columnType = {};
+          typeRecs.forEach((type, field) => {
+            this.$set(this.columnType, field, type);
+          });
```

Comment:
I think we could move this entire function to a utils file or something, so that it's not cluttering this component so heavily. It could instead return `typeRecs`, at which point you'd perform the highlighted operations in this component.

Reply:
Which entire function do you mean? Looking this over, it's a giant call to `Papa.parse()`, with the different callbacks being sort of related to each other in how this operates. I could see abstracting this whole call into a function outside the component definition but I don't see the point of moving just the type heuristic out, or of moving it to a global place like a utils file.

Let me know a bit more about what you're proposing and I can give you fuller thoughts on it.

Reply:
See #131 for my recommendation.

Reply:
Approved #131. So I'll wait for you to merge it, and then we can conduct final reviews on this PR.